File: longyan_mj/evaluator.py
"""Winning-hand evaluation for the first Longyan Mahjong MVP."""

from collections import Counter
from functools import lru_cache
from typing import Dict, Iterable, List, Optional

from .tiles import TERMINALS_AND_HONORS, TILE_INDEX, TILE_ORDER, is_numbered, tile_number


class WinResult:
    """A small value object for a winning hand."""

    def __init__(self, kind: str, label: str, multiplier: int) -> None:
        self.kind = kind
        self.label = label
        self.multiplier = multiplier

    def __repr__(self) -> str:
        return f"WinResult(kind={self.kind!r}, multiplier={self.multiplier!r})"


def count_gold(tiles: Iterable[str], gold_tile: str) -> int:
    return sum(1 for tile in tiles if tile == gold_tile)


def evaluate_win(
    tiles: Iterable[str], gold_tile: Optional[str], open_melds: int = 0
) -> Optional[WinResult]:
    """Return the best supported win result, or None.

    The MVP supports three-gold, seven-pairs, thirteen-orphans and standard
    4-sets-plus-pair hands. Gold tiles are wildcards for hand evaluation.
    """
    hand = list(tiles)
    if gold_tile and count_gold(hand, gold_tile) >= 3:
        return WinResult("three_gold", "三金倒", 3)
    if open_melds == 0 and gold_tile and is_thirteen_orphans(hand, gold_tile):
        return WinResult("thirteen_orphans", "十三幺", 20)
    if open_melds == 0 and gold_tile and is_seven_pairs(hand, gold_tile):
        return WinResult("seven_pairs", "七对", 3)
    if is_standard_win(hand, gold_tile, open_melds=open_melds):
        return WinResult("standard", "自摸", 1)
    return None


def find_youjin_discard(
    tiles: Iterable[str], gold_tile: Optional[str], open_melds: int = 0
) -> Optional[str]:
    """Return a natural pair tile that can be discarded to enter youjin.

    MVP rule: if the hand is a standard winning hand and one gold can serve as
    the pair with a natural tile, discarding that natural tile leaves the gold as
    single pair anchor for single-you.
    """
    if not gold_tile:
        return None
    hand = list(tiles)
    if hand.count(gold_tile) != 1:
        return None
    if not is_standard_win(hand, gold_tile, open_melds=open_melds):
        return None

    counts = Counter(tile for tile in hand if tile != gold_tile)
    for tile in sorted(counts, key=lambda item: TILE_INDEX[item]):
        if counts[tile] >= 1:
            candidate = hand[:]
            candidate.remove(tile)
            if is_standard_win(candidate + [tile], gold_tile, open_melds=open_melds):
                without_gold = [item for item in candidate if item != gold_tile]
                if _can_form_melds(_counts_tuple(Counter(without_gold)), 0, 4 - open_melds):
                    return tile
    return None


def is_seven_pairs(tiles: Iterable[str], gold_tile: str) -> bool:
    hand = list(tiles)
    if len(hand) != 14:
        return False
    wildcards = count_gold(hand, gold_tile)
    counts = Counter(tile for tile in hand if tile != gold_tile)
    odd_count = sum(1 for count in counts.values() if count % 2 == 1)
    return wildcards >= odd_count and (wildcards - odd_count) % 2 == 0


def is_thirteen_orphans(tiles: Iterable[str], gold_tile: str) -> bool:
    hand = list(tiles)
    if len(hand) != 14:
        return False
    wildcards = count_gold(hand, gold_tile)
    counts = Counter(tile for tile in hand if tile != gold_tile)
    if any(tile not in TERMINALS_AND_HONORS for tile in counts):
        return False

    missing = sum(1 for tile in TERMINALS_AND_HONORS if counts.get(tile, 0) == 0)
    if missing > wildcards:
        return False

    remaining_wildcards = wildcards - missing
    has_natural_pair = any(counts.get(tile, 0) >= 2 for tile in TERMINALS_AND_HONORS)
    return has_natural_pair or remaining_wildcards >= 1


def is_standard_win(
    tiles: Iterable[str], gold_tile: Optional[str], open_melds: int = 0
) -> bool:
    hand = list(tiles)
    sets_needed = 4 - open_melds
    if sets_needed < 0:
        return False
    if len(hand) != sets_needed * 3 + 2:
        return False

    wildcards = count_gold(hand, gold_tile) if gold_tile else 0
    counts = Counter(tile for tile in hand if tile != gold_tile)
    counts_tuple = _counts_tuple(counts)

    for index, count in enumerate(counts_tuple):
        if count <= 0:
            continue
        take = min(2, count)
        need = 2 - take
        if need <= wildcards:
            next_counts = list(counts_tuple)
            next_counts[index] -= take
            if _can_form_melds(tuple(next_counts), wildcards - need, sets_needed):
                return True

    if wildcards >= 2 and _can_form_melds(counts_tuple, wildcards - 2, sets_needed):
        return True

    return False


def _counts_tuple(counts: Dict[str, int]) -> tuple:
    return tuple(counts.get(tile, 0) for tile in TILE_ORDER)


@lru_cache(maxsize=None)
def _can_form_melds(counts_tuple: tuple, wildcards: int, sets_needed: int) -> bool:
    if sets_needed == 0:
        return sum(counts_tuple) == 0

    first = _first_non_zero(counts_tuple)
    if first is None:
        return wildcards >= sets_needed * 3

    counts = list(counts_tuple)

    triplet_take = min(3, counts[first])
    triplet_need = 3 - triplet_take
    if triplet_need <= wildcards:
        next_counts = counts[:]
        next_counts[first] -= triplet_take
        if _can_form_melds(tuple(next_counts), wildcards - triplet_need, sets_needed - 1):
            return True

    if _can_make_sequence_from(first):
        indexes = [first, first + 1, first + 2]
        sequence_need = sum(1 for index in indexes if counts[index] == 0)
        if sequence_need <= wildcards:
            next_counts = counts[:]
            for index in indexes:
                if next_counts[index] > 0:
                    next_counts[index] -= 1
            if _can_form_melds(tuple(next_counts), wildcards - sequence_need, sets_needed - 1):
                return True

    return False


def _first_non_zero(counts_tuple: tuple) -> Optional[int]:
    for index, count in enumerate(counts_tuple):
        if count > 0:
            return index
    return None


def _can_make_sequence_from(index: int) -> bool:
    tile = TILE_ORDER[index]
    if not is_numbered(tile):
        return False
    return tile_number(tile) <= 7 and TILE_ORDER[index + 1][:1] == tile[:1]

File: miniprogram/core/evaluator.js
const {
  PLAYABLE_TILES,
  TERMINALS_AND_HONORS,
  TILE_INDEX,
  isNumbered,
  tileNumber,
} = require("./tiles");

function countGold(tiles, goldTile) {
  if (!goldTile) {
    return 0;
  }
  return tiles.filter((tile) => tile === goldTile).length;
}

function evaluateWin(tiles, goldTile, openMelds = 0) {
  const hand = tiles.slice();
  if (goldTile && countGold(hand, goldTile) >= 3) {
    return { kind: "three_gold", label: "三金倒", multiplier: 3 };
  }
  if (openMelds === 0 && goldTile && isThirteenOrphans(hand, goldTile)) {
    return { kind: "thirteen_orphans", label: "十三幺", multiplier: 20 };
  }
  if (openMelds === 0 && goldTile && isSevenPairs(hand, goldTile)) {
    return { kind: "seven_pairs", label: "七对", multiplier: 3 };
  }
  if (isStandardWin(hand, goldTile, openMelds)) {
    return { kind: "standard", label: "自摸", multiplier: 1 };
  }
  return null;
}

function evaluateQiangjin(tiles, goldTile, openMelds = 0) {
  if (!goldTile) {
    return null;
  }
  const hand = tiles.slice();
  if (hand.length === openMelds * 3 + 13) {
    if (isQiangjinShape(hand.concat([goldTile]), goldTile, openMelds)) {
      return { kind: "qiang_jin", label: "抢金", multiplier: 4 };
    }
  }
  if (hand.length === openMelds * 3 + 14) {
    const uniqueTiles = Array.from(new Set(hand)).sort((a, b) => TILE_INDEX[a] - TILE_INDEX[b]);
    for (let index = 0; index < uniqueTiles.length; index += 1) {
      const candidate = hand.slice();
      candidate.splice(candidate.indexOf(uniqueTiles[index]), 1);
      if (isQiangjinShape(candidate.concat([goldTile]), goldTile, openMelds)) {
        return { kind: "qiang_jin", label: "抢金", multiplier: 4 };
      }
    }
  }
  return null;
}

function findYoujinDiscard(tiles, goldTile, openMelds = 0) {
  if (!goldTile) {
    return null;
  }
  const hand = tiles.slice();
  if (countGold(hand, goldTile) >= 3 || !isStandardWin(hand, goldTile, openMelds)) {
    return null;
  }
  const uniqueTiles = Array.from(new Set(hand.filter((tile) => tile !== goldTile)))
    .sort((a, b) => TILE_INDEX[a] - TILE_INDEX[b]);
  for (let index = 0; index < uniqueTiles.length; index += 1) {
    const tile = uniqueTiles[index];
    const candidate = hand.slice();
    candidate.splice(candidate.indexOf(tile), 1);
    if (isYoujinWait(candidate, goldTile, openMelds)) {
      return tile;
    }
  }
  return null;
}

function isQiangjinShape(tiles, goldTile, openMelds) {
  if (openMelds === 0 && isThirteenOrphans(tiles, goldTile)) {
    return true;
  }
  if (openMelds === 0 && isSevenPairs(tiles, goldTile)) {
    return true;
  }
  return isStandardWin(tiles, goldTile, openMelds);
}

function isYoujinWait(tiles, goldTile, openMelds) {
  const setsNeeded = 4 - openMelds;
  if (tiles.length !== setsNeeded * 3 + 1) {
    return false;
  }
  return PLAYABLE_TILES.every((draw) => evaluateWin(tiles.concat([draw]), goldTile, openMelds));
}

function isSevenPairs(tiles, goldTile) {
  if (tiles.length !== 14) {
    return false;
  }
  const wildcards = countGold(tiles, goldTile);
  const counts = countTiles(tiles.filter((tile) => tile !== goldTile));
  const oddCount = Object.keys(counts).filter((tile) => counts[tile] % 2 === 1).length;
  return wildcards >= oddCount && (wildcards - oddCount) % 2 === 0;
}

function isThirteenOrphans(tiles, goldTile) {
  if (tiles.length !== 14) {
    return false;
  }
  const wildcards = countGold(tiles, goldTile);
  const counts = countTiles(tiles.filter((tile) => tile !== goldTile));
  const countKeys = Object.keys(counts);
  if (countKeys.some((tile) => !TERMINALS_AND_HONORS.has(tile))) {
    return false;
  }
  const missing = Array.from(TERMINALS_AND_HONORS).filter((tile) => !counts[tile]).length;
  if (missing > wildcards) {
    return false;
  }
  const remainingWildcards = wildcards - missing;
  const hasNaturalPair = Array.from(TERMINALS_AND_HONORS).some((tile) => counts[tile] >= 2);
  return hasNaturalPair || remainingWildcards >= 1;
}

function isStandardWin(tiles, goldTile, openMelds = 0) {
  const hand = tiles.slice();
  const setsNeeded = 4 - openMelds;
  if (setsNeeded < 0 || hand.length !== setsNeeded * 3 + 2) {
    return false;
  }
  const wildcards = countGold(hand, goldTile);
  const counts = countTiles(hand.filter((tile) => tile !== goldTile));
  const tuple = countsTuple(counts);

  for (let index = 0; index < tuple.length; index += 1) {
    const count = tuple[index];
    if (count <= 0) {
      continue;
    }
    const pairTakes = [];
    if (count >= 2) {
      pairTakes.push(2);
    }
    if (count >= 1) {
      pairTakes.push(1);
    }
    for (let takeIndex = 0; takeIndex < pairTakes.length; takeIndex += 1) {
      const take = pairTakes[takeIndex];
      const need = 2 - take;
      if (need > wildcards) {
        continue;
      }
      const nextTuple = tuple.slice();
      nextTuple[index] -= take;
      if (canFormMelds(nextTuple, wildcards - need, setsNeeded)) {
        return true;
      }
    }
  }

  if (wildcards >= 2 && canFormMelds(tuple, wildcards - 2, setsNeeded)) {
    return true;
  }
  return false;
}

function countTiles(tiles) {
  return tiles.reduce((counts, tile) => {
    counts[tile] = (counts[tile] || 0) + 1;
    return counts;
  }, {});
}

function countsTuple(counts) {
  return PLAYABLE_TILES.map((tile) => counts[tile] || 0);
}

const meldMemo = {};

function canFormMelds(tuple, wildcards, setsNeeded) {
  const key = `${tuple.join(",")}|${wildcards}|${setsNeeded}`;
  if (Object.prototype.hasOwnProperty.call(meldMemo, key)) {
    return meldMemo[key];
  }
  if (setsNeeded === 0) {
    const ok = tuple.reduce((sum, count) => sum + count, 0) === 0;
    meldMemo[key] = ok;
    return ok;
  }
  const first = tuple.findIndex((count) => count > 0);
  if (first < 0) {
    const ok = wildcards >= setsNeeded * 3;
    meldMemo[key] = ok;
    return ok;
  }

  const tripletTake = Math.min(3, tuple[first]);
  const tripletNeed = 3 - tripletTake;
  if (tripletNeed <= wildcards) {
    const nextTuple = tuple.slice();
    nextTuple[first] -= tripletTake;
    if (canFormMelds(nextTuple, wildcards - tripletNeed, setsNeeded - 1)) {
      meldMemo[key] = true;
      return true;
    }
  }

  if (canMakeSequenceFrom(first)) {
    const indexes = [first, first + 1, first + 2];
    const sequenceNeed = indexes.filter((index) => tuple[index] === 0).length;
    if (sequenceNeed <= wildcards) {
      const nextTuple = tuple.slice();
      indexes.forEach((index) => {
        if (nextTuple[index] > 0) {
          nextTuple[index] -= 1;
        }
      });
      if (canFormMelds(nextTuple, wildcards - sequenceNeed, setsNeeded - 1)) {
        meldMemo[key] = true;
        return true;
      }
    }
  }

  meldMemo[key] = false;
  return false;
}

function canMakeSequenceFrom(index) {
  const tile = PLAYABLE_TILES[index];
  if (!isNumbered(tile)) {
    return false;
  }
  return tileNumber(tile) <= 7 && PLAYABLE_TILES[index + 1].slice(0, 1) === tile.slice(0, 1);
}

module.exports = {
  countGold,
  evaluateWin,
  evaluateQiangjin,
  findYoujinDiscard,
  isSevenPairs,
  isThirteenOrphans,
  isStandardWin,
};
